### ISSUE TITLE
latest/edge + default branch name as tag (v2)

### ISSUE THREAD

**Issue by MarcelCoding** [crazy-max/ghaction-docker-meta#55]
### Behaviour
Latest is also created in default branch and not only on tag creation.
Branch name is created in default branch if edge is enabled.

#### Steps to reproduce this issue

1. Enable `type=edge`
2. Enable `type=ref,event=branch`

#### Expected behaviour

Latest only if the run is in a tag.
Branch name if default branch only if edge is disabled.

#### Actual behaviour

see Behaviour

### Configuration

* Repository URL (if public): https://github.com/marcelcoding/luna/
* Build URL (if public): 
  * latest in default branch: https://github.com/MarcelCoding/luna/runs/2228219960?check_suite_focus=true
  * branch name and edge (and latest): https://github.com/MarcelCoding/luna/runs/2227988398?check_suite_focus=true (sorry for the strange tags, I missed the `-` and swapped prefix and suffix)

```yml
      - name: Docker meta
        id: docker_meta
        uses: crazy-max/ghaction-docker-meta@v2
        with:
          images: |
            ${{ env.IMAGE_NAME }}
            ghcr.io/${{ env.IMAGE_NAME }}
          tags: |
            type=edge
            type=ref,event=pr
            type=ref,event=branch
            type=semver,pattern={{version}}
            type=semver,pattern={{major}}.{{minor}}
          flavor: |
            suffix=-${{ matrix.jvm-impl }}
            latest=${{ matrix.jvm-impl == 'hotspot' }}
```

### Logs
first: [first.txt](https://github.com/crazy-max/ghaction-docker-meta/files/6229606/first.txt)
second: [second.txt](https://github.com/crazy-max/ghaction-docker-meta/files/6229618/second.txt)

PS: Is there any way to check all existing tags to decide if the current tag is the latest or only a patch for an older tag?

**Comment by crazy-max** [comment]
@MarcelCoding 

> Latest is also created in default branch and not only on tag creation.

I think that's because you've changed the default behavior of [`latest` flavor](https://github.com/crazy-max/ghaction-docker-meta#flavor-input) (`auto` by default):

```yaml
latest=${{ matrix.jvm-impl == 'hotspot' }}
```

See https://github.com/crazy-max/ghaction-docker-meta#latest-tag

> Branch name is created in default branch if edge is enabled.

You can disable `type=ref,event=branch` if default branch is detected with:

```yaml
type=ref,event=branch,enable=${{ !endsWith(github.ref, github.event.repository.default_branch }}
```

**Comment by MarcelCoding** [comment]
But how does your action know witch is the default flavor?

**Comment by crazy-max** [comment]
If `latest` attribute is not defined in `flavor` input then latest tag will be generated automatically with one of the following conditions:

* on git tag event (`type=ref,event=tag`)
* git tag is a valid semver  (`type=semver,pattern=...`)
* git tag match a pattern (`type=match,pattern=...`)

Valid values for `latest` attribute in flavor is `auto|true|false` as explained in https://github.com/crazy-max/ghaction-docker-meta#flavor-input

In your case I guess you leave it in auto mode. So remove `latest=${{ matrix.jvm-impl == 'hotspot' }}` in flavor input.

**Comment by MarcelCoding** [comment]
But now they is no `edge` tag. Only `edge-openj9` and `edge-hotspot`. (`edge-hotspot` should be also pushed as `edge`) -> How can I define the default flavor, without the issues I mentioned?

**Comment by crazy-max** [comment]
@MarcelCoding 

> But now they is no `edge` tag. Only `edge-openj9` and `edge-hotspot`. (`edge-hotspot` should be also pushed as `edge`) -> How can I define the default flavor, without the issues I mentioned?

That's because you have defined a global `suffix` in `flavor` input so you should add the following tag in `tags` input:

`type=raw,value=edge,enable=${{ matrix.jvm-impl == 'hotspot' && endsWith(github.ref, github.event.repository.default_branch) }}`

**Comment by buchdag** [comment]
@crazy-max I have a very hard time wrapping my head around the way the `latest` flavor is supposed to work (and interact with the tags) from the docs.

If I have a `master` branch on my repo and want the last push event to this branch to be tagged as `:latest` but **not** as `:master`, how should I configure `crazy-max/ghaction-docker-meta@v2` ?

**Comment by crazy-max** [comment]
@buchdag 

> If I have a `master` branch on my repo and want the last push event to this branch to be tagged as `:latest` but **not** as `:master`, how should I configure `crazy-max/ghaction-docker-meta@v2` ?

Looking at your wporkflow then it's a `raw` tag type:

```
tags:
  type=raw,value=latest,enable=${{ endsWith(github.ref, 'master') }}
```

or if 'master' is the default branch:

```
tags:
  type=raw,value=edge,enable=${{ endsWith(github.ref, github.event.repository.default_branch) }}
```

**Comment by buchdag** [comment]
Thanks for your answer, shouldn't

```yaml
tags:
  type=raw,value=edge,enable=${{ endsWith(github.ref, github.event.repository.default_branch) }}
```

rather be

```yaml
tags:
  type=raw,value=latest,enable=${{ endsWith(github.ref, github.event.repository.default_branch) }}
```

?

**Comment by crazy-max** [comment]
oh yes sorry my bad :D

**Comment by buchdag** [comment]
Okay I got it then, no need to actually use the `latest` flavor for this workflow.

**Comment by crazy-max** [comment]
@buchdag 

> Okay I got it then, no need to actually use the `latest` flavor for this workflow.

Yes that's it in your case.

**Comment by buchdag** [comment]
Thanks for you help 👍

**Comment by Flydiverny** [comment]
Hi!
I was having issues with getting tags with the branch name which I didn't want, after some digging I finally figured out that my `type=edge` is acting like a `type=ref,event=branch` in case it doesn't match the branch. This kind of seems like a bug to me, or not what I expected? 😄 (this issue helped me figure it out). I just wanted literal `edge` tags for my default branch but now I also got `branch-name` on branches.

I think I'd expect this place to return if its not matching, or am I missing something @crazy-max ? 😄 
https://github.com/crazy-max/ghaction-docker-meta/blob/f39f06a624f9ca2c5d27758aa88932d443e9eda2/src/meta.ts#L270-L272
